File: routes/mis_proyectos.py
from flask import Blueprint, session, jsonify, request
from services.proyecto_queries import ProyectosQueries
from utils.decorador import requiere_permiso

mis_proyectos = Blueprint('mis_proyectos', __name__)

@mis_proyectos.route('/mis_proyectos/acceso', methods=['GET'])
@requiere_permiso('Acceso Mis Proyectos')
def acceso_mis_proyectos():
    usuario_id = session.get('user_id')

    proyectos = ProyectosQueries.obtener_proyectos(usuario_id)

    if not proyectos:
        return jsonify({
            "message": "Bienvenido a tus proyectos, pero no tienes proyectos asociados.",
            "proyectos": [],
            "opciones": [{"nombre": "Listar proyectos", "endpoint": "/mis_proyectos/listar"}]
        }), 200

    proyectos_json = [
        {"id": proyecto.id, "nombre": proyecto.nombre, "descripcion": proyecto.descripcion}
        for proyecto in proyectos
    ]
    return jsonify({
        "message": "Bienvenido a tus proyectos",
        "proyectos": proyectos_json,
        "opciones": [{"nombre": "Filtrar proyectos", "endpoint": "/mis_proyectos/listar"}]
    }), 200


@mis_proyectos.route('/mis_proyectos/listar', methods=['GET'])
@requiere_permiso('Acceso Mis Proyectos')
def listar_proyectos():
    usuario_id = session.get('user_id')

    nombre = request.args.get('nombre')
    proyectos = ProyectosQueries.obtener_proyectos(usuario_id=usuario_id, nombre=nombre)

    if not nombre and not proyectos:
        proyectos = ProyectosQueries.obtener_todos_proyectos()

    if not proyectos:
        return jsonify({"message": "No se encontraron proyectos"}), 404

    proyectos_json = [
        {"id": proyecto.id, "nombre": proyecto.nombre, "descripcion": proyecto.descripcion}
        for proyecto in proyectos
    ]
    return jsonify(proyectos_json), 200


@mis_proyectos.route('/mis_proyectos/resumen_tickets', methods=['GET'])
@requiere_permiso('Acceso Mis Proyectos')
def resumen_tickets():
    usuario_id = session.get('user_id')
    proyecto_id = request.args.get('proyecto_id')

    try:
        resumen = ProyectosQueries.obtener_resumen_tickets(usuario_id)
        if not resumen:
            return jsonify({"message": "No se encontraron tickets asociados a los proyectos"}), 404

        resumen_json = [
            {
                "proyecto_id": item.fk_proyecto,
                "proyecto_nombre": item.proyecto_nombre,
                "abiertos": item.abiertos,
                "pendientes": item.pendientes,
                "atendidos": item.atendidos,
                "cerrados": item.cerrados,
                "devueltos": item.devueltos,
            } for item in resumen if str(item.fk_proyecto) == proyecto_id
        ]
        if not resumen_json:
            return jsonify({"message": "No se encontraron tickets asociados a los proyectos"}), 404
        
        return jsonify({"resumen": resumen_json}), 200
    except Exception as e:
        return jsonify({"error": f"Error al obtener el resumen de tickets: {str(e)}"}), 500

File: static/mis_proyectos/mis_proyectos.js
const cargarProyectos = async () => {
    try {
        const respuesta = await fetch('/mis_proyectos/acceso');
        const data = await respuesta.json();

        if (respuesta.ok) {
            const proyectosContainer = document.getElementById('proyectos-container');
            proyectosContainer.innerHTML = ''; 

            if (data.proyectos.length > 0) {
                data.proyectos.forEach(proyecto => {
                    const proyectoElemento = document.createElement('div'); 
                    proyectoElemento.classList.add('proyecto-card'); 

                    proyectoElemento.innerHTML = `
                        <h3>${proyecto.nombre}</h3>
                        <p>${proyecto.descripcion}</p>
                        <button class="btn-resumen" data-id="${proyecto.id}">Resumen</button>
                    `;

                    proyectosContainer.appendChild(proyectoElemento); 
                });

                document.querySelectorAll('.btn-resumen').forEach(boton => {
                    boton.addEventListener('click', function() {
                        const proyectoId = this.getAttribute('data-id');
                        cargarResumenTickets(proyectoId);
                    });
                });

            } else {
                proyectosContainer.innerHTML = '<p>No tienes proyectos asociados.</p>';
            }
        } else {
            alert(data.message);
        }
    } catch (error) {
        console.error('Error al cargar proyectos:', error);
    }
};



const filtrarProyectos = async () => {
    const nombreFiltro = document.getElementById('input-filtro').value;
    const url = `/mis_proyectos/listar?nombre=${nombreFiltro}`; 

    try {
        const respuesta = await fetch(url);
        const data = await respuesta.json();

        if (respuesta.ok) {
            const proyectosContainer = document.getElementById('proyectos-container');
            proyectosContainer.innerHTML = '';

            data.forEach(proyecto => {
                const proyectoElemento = `
                    <div class="proyecto-card">
                        <h3>${proyecto.nombre}</h3>
                        <p>${proyecto.descripcion}</p>
                    </div>
                `;
                proyectosContainer.innerHTML += proyectoElemento;
            });
        } else {
            alert(data.message);
        }
    } catch (error) {
        console.error('Error al filtrar proyectos:', error);
    }
};


const cargarResumenTickets = async (proyectoId) => {
    try {
        const respuesta = await fetch(`/mis_proyectos/resumen_tickets?proyecto_id=${proyectoId}`);
        const data = await respuesta.json();

        if (respuesta.ok) {
            const resumenContainer = document.getElementById('resumen-tickets-container');
            resumenContainer.innerHTML = ''; 

            data.resumen.forEach(item => {
                if (item.proyecto_id == proyectoId) { 
                    const resumenElemento = `
                        <div class="resumen-card">
                            <h3>${item.proyecto_nombre}</h3>
                            <p>Abiertos: ${item.abiertos}</p>
                            <p>Pendientes: ${item.pendientes}</p>
                            <p>Atendidos: ${item.atendidos}</p>
                            <p>Cerrados: ${item.cerrados}</p>
                            <p>Devueltos: ${item.devueltos}</p>
                        </div>
                    `;
                    resumenContainer.innerHTML += resumenElemento;
                }
            });

        } else {
            alert(data.message);
        }
    } catch (error) {
        console.error('Error al cargar resumen tickets:', error);
    }
};